### ISSUE TITLE
Need bypass CRL if customer have internal CA for oneview

### ISSUE THREAD

**Issue by bluesky255** [HewlettPackard/POSH-HPOneView#248]
I have internal CA issued ssl to oneview. cert is valid in IE (root cert imported) but cannot use in powershell Connect-HPOVMgmt . Is there anyway to bypass SSL check?

Please fill in as much information as possible to help resolve your issue.

### Expected Behavior
can login

### Actual Behavior
```PowerShell
PS Cert:\LOCALMACHINE> Connect-HPOVMgmt -Hostname hpeoneview.hpe.local -UserName Administrator -AuthLoginDomain Local -LoginAcknowledge -Password 



Certificate             : [Subject]
                            CN=hpeoneview.hpe.local, OU=HK, O=Hewlett-Packard, L=Hong Kong, S=HK, C=HK
                          
                          [Issuer]
                            CN=hpe-HPEDC01-CA, DC=hpe, DC=local
                          
                          [Serial Number]
                            1D000000082D9B9CE488C364EB000000000008
                          
                          [Not Before]
                            7/11/2016 2:03:15 PM
                          
                          [Not After]
                            7/11/2018 2:03:15 PM
                          
                          [Thumbprint]
                            B7E83201B609A451ED9E79DDD7E0951C455FF1C4
                          
Issuer                  : CN=hpe-HPEDC01-CA, DC=hpe, DC=local
Subject                 : CN=hpeoneview.hpe.local, OU=HK, O=Hewlett-Packard, L=Hong Kong, S=HK, C=HK
SubjectAlternativeNames : DNS Name=hpeoneview.hpe.local DNS Name=hpeoneview IP Address={IP} DNS Name= {IP}
CertificateIsValid      : False
ErrorInformation        : RevocationStatusUnknown OfflineRevocation

Connect-HPOVMgmt : The underlying connection was closed: Could not establish trust relationship for the SSL/TLS secure channel.
At line:1 char:1
+ Connect-HPOVMgmt -Hostname hpeoneview.hpe.local -UserName Administrator -AuthLog ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Connect-HPOVMgmt], WebException
    + FullyQualifiedErrorId : WebException,Connect-HPOVMgmt
```

### Version Information
HPE OneView PowerShell Library Version (`Get-HPOVVersion`):

HPE OneView Appliance Version:  
Output from `$PSVersionTable` on your Windows Host:
```PowerShell
PS Cert:\LOCALMACHINE> $PSVersionTable

Name                           Value                                                                                                 
----                           -----                                                                                                 
PSVersion                      4.0                                                                                                   
WSManStackVersion              3.0                                                                                                   
SerializationVersion           1.1.0.1                                                                                               
CLRVersion                     4.0.30319.42000                                                                                       
BuildVersion                   6.3.9600.16406                                                                                        
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0}                                                                                  
PSRemotingProtocolVersion      2.2      

PS Cert:\LOCALMACHINE> Get-ChildItem -path Cert:\CurrentUser\Root | ? ThumbPrint -eq  B7E83201B609A451ED9E79DDD7E0951C455FF1C4


    Directory: Microsoft.PowerShell.Security\Certificate::CurrentUser\Root


Thumbprint                                Subject                                                                                    
----------                                -------                                                                                    
B7E83201B609A451ED9E79DDD7E0951C455FF1C4  CN=hpeoneview.hpe.local, OU=HK, O=Hewlett-Packard, L=Hong Kong, S=HK, C=HK                 

```

**Comment by ChrisLynchHPE** [comment]
Can you get the output from:

```PowerShell
dir Cert:\CurrentUser\Root | ? Subject -match 'CN=hpe-HPEDC01-CA' | fl *
```

What is strange is you have trusted the appliance certificate, yet Windows believes it is untrusted.

**Comment by ChrisLynchHPE** [comment]
And I cannot seem to reproduce this issue.  I have an internal issuing CA, which is trusted by my PC, am able to connect to my appliance with either a Self-Signed Cert (even went so far as to trust the certificate) or issued Cert without incident.

Can you remove the certificate with the `B7E83201B609A451ED9E79DDD7E0951C455FF1C4` thumbprint, and try using `Import-HPOVSSLCertificate` Cmdlet and try again, even though your PC should trust the issuing CA in the trusted root store?

**Comment by bluesky255** [comment]
Yes, it is strange.

PS C:\Users\chengwi> dir Cert:\CurrentUser\Root | ? Subject -match 'CN=hpe-HPEDC01-CA' | fl *


PSPath               : Microsoft.PowerShell.Security\Certificate::CurrentUser\Root\33DD93088DA61EB15A9D455615F37D4F2E0F
                       23AE
PSParentPath         : Microsoft.PowerShell.Security\Certificate::CurrentUser\Root
PSChildName          : 33DD93088DA61EB15A9D455615F37D4F2E0F23AE
PSDrive              : Cert
PSProvider           : Microsoft.PowerShell.Security\Certificate
PSIsContainer        : False
EnhancedKeyUsageList : {}
DnsNameList          : {}
SendAsTrustedIssuer  : False
Archived             : False
Extensions           : {System.Security.Cryptography.Oid, System.Security.Cryptography.Oid,
                       System.Security.Cryptography.Oid, System.Security.Cryptography.Oid}
FriendlyName         :
IssuerName           : System.Security.Cryptography.X509Certificates.X500DistinguishedName
NotAfter             : 7/11/2026 9:16:57 AM
NotBefore            : 7/11/2016 9:06:57 AM
HasPrivateKey        : False
PrivateKey           :
PublicKey            : System.Security.Cryptography.X509Certificates.PublicKey
RawData              : {48, 130, 3, 101...}
SerialNumber         : 47E8823A8804B399407A39BE9EEBE3E9
SubjectName          : System.Security.Cryptography.X509Certificates.X500DistinguishedName
SignatureAlgorithm   : System.Security.Cryptography.Oid
Thumbprint           : 33DD93088DA61EB15A9D455615F37D4F2E0F23AE
Version              : 3
Handle               : 473116272
Issuer               : CN=hpe-HPEDC01-CA, DC=hpe, DC=local
Subject              : CN=hpe-HPEDC01-CA, DC=hpe, DC=local

**Comment by bluesky255** [comment]
I can import the cert but may I know how to remove thumbprint?

PS Cert:\LOCALMACHINE> Import-HPOVSslCertificate -ApplianceConnection hpeoneview.hpe.local -Verbose
VERBOSE: [IMPORT-HPOVSSLCERTIFICATE] Bound PS Parameters: 
Key                                                                Value                                                             
---                                                                -----                                                             
ApplianceConnection                                                hpeoneview.hpe.local                                              
Verbose                                                            True                                                              



VERBOSE: [IMPORT-HPOVSSLCERTIFICATE] Called from: <ScriptBlock>
VERBOSE: [IMPORT-HPOVSSLCERTIFICATE] Processing Appliance 'hpeoneview.hpe.local' (of 1)
VERBOSE: [IMPORT-HPOVSSLCERTIFICATE] Getting response
VERBOSE: Closing response connection
VERBOSE: [IMPORT-HPOVSSLCERTIFICATE] Attempting to add cert to store
VERBOSE: [IMPORT-HPOVSSLCERTIFICATE] Cert added successfully
WARNING: Please note that the Subject Alternate Name (SAN) must match that of the Appliance hostname you use to connect to your applia
nce.  If it does not, an SSL connection failure will occur.  When creating a CSR on the appliance, make sure to include the additional
 FQDN and IP address(es) in the Alternative Name field.

PS Cert:\LOCALMACHINE>

**Comment by ChrisLynchHPE** [comment]
If you want to remove a certificate using PowerShell, use the following:

```PowerShell
PS C:\> Get-ChildItem -path Cert:\CurrentUser\Root | ? ThumbPrint -eq  B7E83201B609A451ED9E79DDD7E0951C455FF1C4 | Remove-Item
```
You could also use the Certifcates MMC snap-in.

**Comment by bluesky255** [comment]
Issue can't fix by remove the certificate. 

VERBOSE: [RESTCLIENT] Building new [System.Net.HttpWebRequest] object for GET https://hpeoneview.hpe.local/rest/version
VERBOSE: [SEND-HPOVREQUEST] Request: GET https://hpeoneview.hpe.local/rest/version

VERBOSE: [SEND-HPOVREQUEST] Net.WebException Error caught
VERBOSE: [SEND-HPOVREQUEST] The underlying connection was closed: Could not establish trust relationship for the SSL/TLS secure channel.
VERBOSE: [SEND-HPOVREQUEST] Cleaning up HttpWebRequest
Connect-HPOVMgmt : The underlying connection was closed: Could not establish trust relationship for the SSL/TLS secure channel.
At line:1 char:1
+ Connect-HPOVMgmt -Hostname hpeoneview.hpe.local -UserName Administrator -AuthLog ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Connect-HPOVMgmt], WebException
    + FullyQualifiedErrorId : WebException,Connect-HPOVMgmt

=======================================================
PS C:\Users\chengwi> $url = "https://hpeoneview.hpe.local/rest/version"
PS C:\Users\chengwi>
PS C:\Users\chengwi> $request = [System.Net.httpwebrequest]::Create($url)
PS C:\Users\chengwi> $request.Method="Get"
PS C:\Users\chengwi> $response = $request.GetResponse()
PS C:\Users\chengwi> $requestStream = $response.GetResponseStream()
PS C:\Users\chengwi> $readStream = New-Object System.IO.StreamReader $requestStream
PS C:\Users\chengwi> $data=$readStream.ReadToEnd()
PS C:\Users\chengwi> if($response.ContentType -match "application/xml") {
>>     $results = [xml]$data
>> } elseif($response.ContentType -match "application/json") {
>>     $results = $data | ConvertFrom-Json
>> } else {
>>     try {
>>         $results = [xml]$data
>>     } catch {
>>         $results = $data | ConvertFrom-Json
>>     }
>> }
>> $results
>>

                                             currentVersion                                              minimumVersion
                                             --------------                                              --------------
                                                        300                                                           1


PS C:\Users\chengwi>

**Comment by ChrisLynchHPE** [comment]
Again, I can't seem to reproduce this behavior.  The certificate is being reported that it is not valid, which is what the `CertificateIsValid` property of the cert is showing.  Is the `hpe-HPEDC01-CA` issuing CA accessible from your PC?  The Cert Chain is validated to check if the cert is revoked.  If the issuing CA, or the CRL location is not available, then the cert chain cannot be validated, thus is not "valid".

So, you need to figure out what is going on with the CRL location, which is defined in the Certificate Template on the issuing CA.

In the meantime, you can use the `$Global:IgnoreCertPolicy = $true` method to bypass the SSL certificate check.  It is not something anyone should use, and can expose a security risk.  It is simply a workaround until you can fix the issue with the CRL.

**Comment by bluesky255** [comment]
Yes, thanks. it work for $Global:IgnoreCertPolicy = $true.  Thanks for your help. Cheers!

**Comment by ChrisLynchHPE** [comment]
Please know that using IgnoreCertPolicy can be a security vulnerability.  All SSL checks will be bypassed, and if there is a man-in-the-middle attacker, you will never know.

**Comment by bluesky255** [comment]
Noted and thanks.  It is my lab and demo use only.

**Comment by ChrisLynchHPE** [comment]
This is fixed in [Release 3.0.1264.2772](https://github.com/HewlettPackard/POSH-HPOneView/releases/tag/v3.0.1264.2772).
